src/models/migrations/MigrationsStore.ts

```diff
@@ -2,26 +2,68 @@
 
 import _ from 'lodash';
 import appData from '../appData';
+const log = require('../log');
 
 const DOCUMENT_ID = 'migrations';
 
 type DoneCallback = (err: unknown|null, state?: unknown) => void;
 
+const safeMigrationMsg = 'safeMigration is true, we do not allow empty migrations to be saved. ' +
+    'If this is new project, feel free to enable it';
+
+function checkValidMigrations(objWithMigrations: { migrations: any; }) {
+    if (
+        !objWithMigrations ||
+        _.isNull(objWithMigrations.migrations) ||
+        _.isEmpty(objWithMigrations.migrations) ||
+        !_.isArray(objWithMigrations.migrations)) {
+        log.error(safeMigrationMsg);
+        throw new Error(safeMigrationMsg);
+    }
+}
+
 class MigrationsStore {
+    safeMigration: boolean;
+
+    constructor(safeMigration: boolean) {
+        this.safeMigration = safeMigration;
+    }
 
     save (set: Object, done: DoneCallback) {
-        appData.updateDocument(DOCUMENT_ID, { $set: _.pick(set, ['lastRun', 'migrations']) }, true)
-            .then((state) => { done(null, state); }, done);
+        log.info('Migration SET', set)
+        const objToUpdate: any = _.pick(set, ['lastRun', 'migrations']);
+        if (this.safeMigration) {
+            checkValidMigrations(objToUpdate);
+
+            const numOfValidMigrations = objToUpdate.migrations
+                .reduce((prev: number, cur: { timestamp: any; }) => cur.timestamp? prev + 1 : prev, 0);
+
+            // If all migrations are renewed=without timestamp (beside the first one that is trying to be applied)
+            // There can be something wrong
+            if (objToUpdate.migrations.length >= 3 && numOfValidMigrations <= 1) {
+                log.error(safeMigrationMsg);
+                throw new Error(safeMigrationMsg);
+            }
+        }
+
+        appData.updateDocument(DOCUMENT_ID, { $set: objToUpdate }, true)
+            .then((state) => { done(null, state); }, done)
+            .catch(done);
     }
 
     load (done: DoneCallback) {
         appData.getDocument(DOCUMENT_ID)
             .then((state) => {
+                if (this.safeMigration) {
+                    checkValidMigrations(state);
+                }
+
                 done(null, state || {
                     lastRun: null,
                     migrations: []
                 });
-            }, done);
+            }, done)
+            .catch(done);
```

Comment:
the done is already passed as the second .then() parameter, remove the original one

Reply:
I simulated the error on locahost and got this

```
BulkWriteError: E11000 duplicate key error collection: jsRouter.tariffs index: _id_ dup key: { _id: ObjectId('5c617e3f3035c219e7781337') }
    at OrderedBulkOperation.handleWriteError (C:\storyous\src\common-utils\node_modules\mongodb\lib\bulk\common.js:1355:9)
    at resultHandler (C:\storyous\src\common-utils\node_modules\mongodb\lib\bulk\common.js:579:23)
    at handler (C:\storyous\src\common-utils\node_modules\mongodb\lib\core\sdam\topology.js:961:24)
    at fn (C:\storyous\src\common-utils\node_modules\mongodb\lib\cmap\connection_pool.js:351:13)
    at handleOperationResult (C:\storyous\src\common-utils\node_modules\mongodb\lib\core\sdam\server.js:567:5)
    at MessageStream.messageHandler (C:\storyous\src\common-utils\node_modules\mongodb\lib\cmap\connection.js:308:5)
    at MessageStream.emit (events.js:198:13)
    at MessageStream.EventEmitter.emit (domain.js:448:20)
    at processIncomingData (C:\storyous\src\common-utils\node_modules\mongodb\lib\cmap\message_stream.js:144:12)
    at MessageStream._write (C:\storyous\src\common-utils\node_modules\mongodb\lib\cmap\message_stream.js:42:5)
    at doWrite (_stream_writable.js:415:12)
    at writeOrBuffer (_stream_writable.js:399:5)
    at MessageStream.Writable.write (_stream_writable.js:299:11)
    at Socket.ondata (_stream_readable.js:710:20)
    at Socket.emit (events.js:198:13)
    at Socket.EventEmitter.emit (domain.js:448:20)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at Socket.Readable.push (_stream_readable.js:224:10)
    at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
From previous event:
    at Migration.up (C:\storyous\src\pay-js-router\app\migrations\1549892168289-tariffs-for-andorra.js:48:5)
    at next (C:\storyous\src\common-utils\node_modules\migrate\lib\migrate.js:35:43)
    at C:\storyous\src\common-utils\node_modules\migrate\lib\migrate.js:71:9
    at store.save (C:\storyous\src\common-utils\node_modules\migrate\lib\set.js:84:5)
    at appData_1.default.updateDocument.then (C:\storyous\src\common-utils\lib\models\migrations\MigrationsStore.js:29:32)
    at process._tickCallback (internal/process/next_tick.js:61:11)
```

Is "from previous event" part new default in node v12 or v14?
Because we are using v10 for IS, maybe its not that powerful for stacktrace